tests/test_unit.py

```diff
@@ -2366,6 +2366,182 @@ def testParseResultsWithNamedTuple(self):
+            result, ["spam", "eggs"], msg="issue with ParseResults before clear()"
+        )
+
+        result.clear()
+
+        print(result.dump())
+        self.assertParseResultsEquals(
+            result, expected_list=[], msg="issue with ParseResults.clear() as list"
+        )
+        self.assertParseResultsEquals(
+            result, expected_dict={}, msg="issue with ParseResults.clear() as dict"
+        )
+
+    def testParseResultsExtendWithString(self):
+        """test ParseResults.extend() with input of type str"""
+
+        # use a parse action to append the reverse of the matched strings to make a palindrome
+        def make_palindrome(tokens):
+            tokens.extend(reversed([t[::-1] for t in tokens]))
+            return "".join(tokens)
+
+        tst = "abc def ghi"
+        expr = pp.OneOrMore(pp.Word(pp.alphas))
+        result = expr.addParseAction(make_palindrome).parseString(tst)
+        print(result.dump())
+
+        expected = ["abcdefghiihgfedcba"]
+        self.assertParseResultsEquals(
+            result, expected, msg="issue with ParseResults.extend(str)"
+        )
+
+    def testParseResultsExtendWithParseResults(self):
+        """test ParseResults.extend() with input of type ParseResults"""
+
+        expr = pp.OneOrMore(pp.Word(pp.alphas))
+        result1 = expr.parseString("spam eggs")
+        result2 = expr.parseString("foo bar")
+
+        result1.extend(result2)
+        print(result1.dump())
+        expected = ["spam", "eggs", "foo", "bar"]
+        self.assertParseResultsEquals(
+            result1, expected, msg="issue with ParseResults.extend(ParseResults)"
+        )
+
+    def testParseResultsFromDict(self):
+        """test helper classmethod ParseResults.from_dict()"""
+
+        dict = {
+            "first": "123",
+            "second": 456,
+            "third": {"threeStr": "789", "threeInt": 789},
+        }
+        name = "trios"
+        result = pp.ParseResults.from_dict(dict, name=name)
+
+        print(result.dump())
+        expected = {name: dict}
+        self.assertParseResultsEquals(
+            result,
+            expected_dict=expected,
+            msg="issue creating ParseResults.from _dict()",
+        )
+
+    def testParseResultsDir(self):
+        """test dir(ParseResults)"""
+
+        dict = {"first": "123", "second": "456", "third": "789"}
+        name = "trios"
+        result = pp.ParseResults.from_dict(dict, name=name)
+        dir_result = dir(result)
+
+        print(dir_result)
+        self.assertIn(
+            name, dir_result, msg="name value wasn't returned by dir(ParseResults)"
+        )
+        self.assertIn(
+            "asList", dir_result, msg="asList was not returned by dir(ParseResults)"
+        )
+
+    def testParseResultsInsert(self):
+        """test ParseResults.insert() with named tokens"""
+
+        # use a parse action to insert the parse location in the front of the parsed results
```

Comment:
This test seems a little light, only does one insert at location 0. Let's create a ParseResults by parsing "A B C D E F G H I J" into separate tokens, and create a list with the same contents. Then do 10 random inserts of "a b c d e f g h i j".split() at locations between -5 and 5 to both the ParseResults and the list. Then print them both and compare if they are equal.

    import pyparsing as pp
    from random import randint

    result = pp.Word(pp.alphas)[...].parseString("A B C D E F G H I J")
    compare_list = result.asList()

    print(result)
    print(compare_list)

    for s in "abcdefghij":
        index = randint(-5, 5)
        result.insert(index, s)
        compare_list.insert(index, s)

    print(result)
    print(compare_list)